### DISCUSSION TITLE
MVP Scope

### DISCUSSION THREAD

**Discussion by spicyfalafel** [Aidbox/viewdefinitions#118] [General]
The goal of ViewDefinition Designer is to provide better UX to create ViewDefinition, compared to plain text editing.
UX instantly gives feedback of what the current VD will do. 

MVP is a version of a product with just enough features to be usable by early *customers* who can then provide feedback for future product development.
By customers we mean Aidbox team to evaluate.
The goal of MVP is to make a decision about the UX of the ViewDefinition form.

# [MVP scope](https://github.com/Aidbox/viewdefinitions/milestone/1)

1. Form page
- [x] user can edit existing ViewDefinition resource in form
- [x] user can get data from db of current VD by clicking the button "Run"
- [x] **styling is good in our opinion** (#45, #46, #47, #48, #36)
- [x] user can choose which elements are shown in the form by hitting "options" icon at every "type" of SELECT and also on the from itself (root elements) (#37)
- [x] user can save VD resource to the server by clicking "create" VD button in VD Form (for "edit" and "create" mode) (#41)
- [x] user can view and download ViewDefinition resource represented in YAML & JSON in read-only mode and cannot modify it (#26)
- [x] user can view SQL result in the tab (#42)
- [x] when VD is saved by clicking "save" button, show "saved" message (#51)
- [x] if "get data from db by VD resource" request fails, show the error message (#51)

2. List of ViewDefinitions page
- [x] user can see which ViewDefinitions resources are already present in connected FHIR server
- [x] user can search for ViewDefinitions
- [x] user can delete VD resource from the server (only external server)
- [x] create an opinionated collection of VDs from SoF tests playground website which is good start point to understand possible ViewDefinitions #77

3. Settings page 
- [x] user can do nothing with the settings and use preconfigured aidbox instance in read-only mode
- [x] user can create and save any number of configurations (see VANYA viewer server list)
- [x] user can choose and configure:
  - [x] URL to get data from (/metadata, /ViewDefinition)
  - [x] any custom headers for every request e.g. to use basic authentication
- [x] user can see FHIR version of connected server
- [x] user understands if server is "connected" successfully and see errors if something went wrong

**Comment by Leningradez** [comment]
### Suggestions

ViewDefinition Designer aims to provide better UX to create and debug ViewDefinitions than JSON/YAML editing. 
Better UX is achieved through the following features:
* visual editor with the predefined structure of ViewDefinition resource,
* real-time representation of results based on sandbox data and debug API,
* basic autocomplete feature for FHIRpath expressions ([path selection only](https://hl7.org/fhirpath/#path-selection))
* education tooltips with element definitions

@spicyfalafel
